### PR TITLE
Adjoint metric tensor uses `devices.qubit` instead of `DefaultQubit` private methods

### PR DESCRIPTION
**Pull request by albi3ro**
We are preparing to switch `pennylane.devices.DefaultQubit` for `pennylane.devices.experimental.DefaultQubit2`. The `adjoint_metric_tensor` transform currently uses private methods of `DefaultQubit` that are going away.

This PR merely substitutes those private methods for functions form the `pennylane.devices.qubit` module.

### CODE REVIEW

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -158,6 +158,9 @@ array([False, False])
 * The experimental `DefaultQubit2` device now supports computing VJPs and JVPs using the adjoint method.
   [(#4374)](https://github.com/PennyLaneAI/pennylane/pull/4374)
 
+* `qml.transforms.adjoint_metric_tensor` now uses the simulation tools in `pennylane.devices.qubit` instead of
+  private methods of `pennylane.devices.DefaultQubit`.
```
Comment (albi3ro) [inline doc/releases/changelog-dev.md:162]:
```suggestion
  private methods of `pennylane.devices.DefaultQubit`.
  [(#4456)](https://github.com/PennyLaneAI/pennylane/pull/4456)
```

File: pennylane/transforms/adjoint_metric_tensor.py
Comment (timmysilv) [inline pennylane/transforms/adjoint_metric_tensor.py:1]: iirc dots in python cost a fair bit of performance. can you import things from `pennylane.devices.qubit` and use those in this file?

Comment (albi3ro) [reply pennylane/transforms/adjoint_metric_tensor.py:1]: circular imports.  Maybe I could import them inside the relevant function?

File: pennylane/transforms/adjoint_metric_tensor.py [hunk 2]
```diff
@@ -20,41 +20,20 @@
...
+    return qml.QubitUnitary(real_mat, generator.wires)
 
-    apply_op = qml.adjoint(op) if invert else op
-    state = device._apply_operation(state, apply_op)
 
-    return state
+def _reshape_real_imag(state, dim):
+    real_part = qml.math.reshape(qml.math.real(state), (dim,))
+    imag_part = qml.math.reshape(qml.math.imag(state), (dim,))
+    return real_part, imag_part
```
Comment (dwierichs) [inline pennylane/transforms/adjoint_metric_tensor.py:36]: Probably not important, but could also do with one reshape? :thinking: 
```suggestion
    state = qml.math.reshape(state, (dim,))
    return qml.math.real(state), qml.math.imag(state)
```

File: pennylane/transforms/adjoint_metric_tensor.py [hunk 3]
```diff
@@ -157,78 +136,92 @@ def circuit(weights):
...
     # pylint: disable=protected-access
-    if device.shots is not None:
+    if tape.shots:
         raise ValueError(
             "The adjoint method for the metric tensor is only implemented for shots=None"
         )
+
+    if set(tape.wires) != set(range(tape.num_wires)):
+        wire_map = {w: i for i, w in enumerate(tape.wires)}
+        tape = qml.map_wires(tape, wire_map)
```
Comment (dwierichs) [inline pennylane/transforms/adjoint_metric_tensor.py:156]: Is this extending functionality or is it just necessary for compatibility with the new device methods? :)

File: pennylane/transforms/adjoint_metric_tensor.py [hunk 4]
```diff
@@ -157,78 +136,92 @@ def circuit(weights):
...
+        phi_real, phi_imag = _reshape_real_imag(phi, dim)
         diag_value = prefactor_1**2 * (
             qml.math.dot(phi_real, phi_real) + qml.math.dot(phi_imag, phi_imag)
         )
-        L = qml.math.scatter_element_add(L, (j, j), diag_value)
 
         lam = psi * 1.0
+        lam_real, lam_imag = _reshape_real_imag(lam, dim)
         lam_real = qml.math.reshape(qml.math.real(lam), (dim,))
         lam_imag = qml.math.reshape(qml.math.imag(lam), (dim,))
```
Comment (dwierichs) [inline pennylane/transforms/adjoint_metric_tensor.py:193]:
```suggestion
```

File: pennylane/transforms/adjoint_metric_tensor.py [hunk 5]
```diff
@@ -157,78 +136,92 @@ def circuit(weights):
...
 
         # this entry is missing a factor of 1j
         value = prefactor_1 * (qml.math.dot(lam_real, phi_real) + qml.math.dot(lam_imag, phi_imag))
+        L = qml.math.scatter_element_add(L, (j, j), diag_value)
         T = qml.math.scatter_element_add(T, (j,), value)
 
+        lam = psi * 1.0
+
+        ###
+
```
Comment (dwierichs) [inline pennylane/transforms/adjoint_metric_tensor.py:203]: Why was this added?

File: pennylane/transforms/adjoint_metric_tensor.py [hunk 6]
```diff
@@ -157,78 +136,92 @@ def circuit(weights):
...
-        L = qml.math.scatter_element_add(L, (j, j), diag_value)
 
         lam = psi * 1.0
+        lam_real, lam_imag = _reshape_real_imag(lam, dim)
         lam_real = qml.math.reshape(qml.math.real(lam), (dim,))
         lam_imag = qml.math.reshape(qml.math.imag(lam), (dim,))
 
         # this entry is missing a factor of 1j
         value = prefactor_1 * (qml.math.dot(lam_real, phi_real) + qml.math.dot(lam_imag, phi_imag))
+        L = qml.math.scatter_element_add(L, (j, j), diag_value)
```
Comment (dwierichs) [inline pennylane/transforms/adjoint_metric_tensor.py:197]: Wouldn't it be good to keep this close to where `diag_value` is computed? :thinking:

**Review by dwierichs — COMMENTED** [review]
Very nice to clean this up and remove the custom `_apply_operations` method :) :tada: 

LGTM, just some very minor points, including an addition of 4 lines that can be skipped, I think (?)

File: pennylane/transforms/adjoint_metric_tensor.py (hunk 2 above)
Comment (albi3ro) [reply pennylane/transforms/adjoint_metric_tensor.py:36]: Probably. Was just copy and pasting what already there.

File: pennylane/transforms/adjoint_metric_tensor.py (hunk 3 above)
Comment (albi3ro) [reply pennylane/transforms/adjoint_metric_tensor.py:155]: The functions in `devices.qubit` assume that the wire labels have been mapped to statevector dimensions.

File: pennylane/transforms/adjoint_metric_tensor.py (hunk 6 above)
Comment (albi3ro) [reply pennylane/transforms/adjoint_metric_tensor.py:194]: Should be fine. I think this is just a byproduct of trying and failing to extract more helper methods.

File: pennylane/transforms/adjoint_metric_tensor.py
Comment (timmysilv) [reply pennylane/transforms/adjoint_metric_tensor.py:1]: I think I'd rather just import them all at the top in that case

File: pennylane/qnode.py [hunk 7]
```diff
@@ -850,6 +850,8 @@ def construct(self, args, kwargs):  # pylint: disable=too-many-branches
         self._tape = make_qscript(self.func, shots)(*args, **kwargs)
         self._qfunc_output = self.tape._qfunc_output
 
+        print("tape in construct", self._tape.wires)
+
```
Comment (dwierichs) [inline pennylane/qnode.py:854]:
```suggestion
```

File: pennylane/transforms/adjoint_metric_tensor.py [hunk 8]
```diff
@@ -20,41 +20,19 @@
...
-
-    if isinstance(op, qml.QubitStateVector):
-        if invert:
-            raise ValueError("Can't invert state preparation.")
-        device._apply_state_vector(op.parameters[0], op.wires)
-        return device._state
+def _get_op_for_generator(generator, like_real):
+    mat = qml.matrix(generator)
+    real_mat = qml.math.convert_like(mat, like_real)
+    return qml.QubitUnitary(real_mat, generator.wires)
```
Comment (dwierichs) [inline pennylane/transforms/adjoint_metric_tensor.py:30]: I got confused while reading this because  `real_mat` is not actually a real matrix.

I think this is due to poor naming (that might be on me, sorry) of `like_real`. While `like_real` is a real-valued object, `convert_like`, which only changes the interface, does not cast to reals.
Maybe, at least within this 4-line function, we could switch to `like_real->like` and `real_mat->mat_interface` or so?

File: pennylane/transforms/adjoint_metric_tensor.py [hunk 9]
```diff
@@ -157,56 +135,59 @@ def circuit(weights):
...
 
     # Divide all operations of a tape into trainable operations and blocks
     # of untrainable operations after each trainable one.
     trainable_operations, group_after_trainable_op = _group_operations(tape)
 
-    dim = 2**device.num_wires
+    dim = 2**tape.num_wires
     # generate and extract initial state
-    psi = device._create_basis_state(0)
+    prep = tape[0] if len(tape) > 0 and isinstance(tape[0], qml.operation.StatePrep) else None
```
Comment (timmysilv) [inline pennylane/transforms/adjoint_metric_tensor.py:163]: just commenting so it doesn't cause a conflict - Jay's change might beat this one, and if it doesn't someone should tell him :)

**Review by timmysilv — APPROVED** [review]
very nice stuff

File: pennylane/transforms/adjoint_metric_tensor.py (hunk 9 above)
Comment (albi3ro) [reply pennylane/transforms/adjoint_metric_tensor.py:163]: @Jaybsoni

Comment (albi3ro) [reply pennylane/transforms/adjoint_metric_tensor.py:163]: thansk for catching that.

File: pennylane/transforms/adjoint_metric_tensor.py (hunk 8 above)
Comment (albi3ro) [reply pennylane/transforms/adjoint_metric_tensor.py:30]: Do we actually even need to convert it to the interface? Without this conversion, we could just do:
```
state = apply_operation(generator, state)
```

But I don't understand this history of why we have to convert, so I left it be.

Comment (albi3ro) [reply pennylane/transforms/adjoint_metric_tensor.py]: Yeah I'm just now using `apply_operation` with the generator`.

**Review by dwierichs — APPROVED** [review]
Thanks @albi3ro :) :100:
